flocker/control/_model.py

```diff
@@ -504,6 +504,126 @@ def get_node(deployment, uuid, **defaults):
     return get_node
 
 
+class LeaseAcquisitionError(ValueError):
+    """
+    Exception raised when a ``Lease`` cannot be acquired.
+    """
+    def __init__(self, message=u""):
+        super(LeaseAcquisitionError, self).__init__(message)
+
+
+class LeaseReleaseError(ValueError):
+    """
+    Exception raised when a ``Lease`` cannot be released.
+    """
+    def __init__(self, message=u""):
+        super(LeaseReleaseError, self).__init__(message)
+
+
+class Lease(PClass):
+    """
+    A lease mapping a dataset to a node, with optional expiry.
+
+    :ivar UUID dataset_id: The dataset this lease represents.
+    :ivar UUID node_id: The node holding this lease.
+    :ivar datetime expiration: The ``datetime`` at which this lease expires.
+    """
+    dataset_id = field(type=UUID)
+    node_id = field(type=UUID)
+    expiration = field(
+        type=(datetime, type(None)), mandatory=True, initial=None
+    )
+
+
+class Leases(CheckedPMap):
+    """
+    A representation of all leases in a cluster, mapped by dataset id.
+    """
+    __key_type__ = UUID
+    __value_type__ = Lease
+
+    def __invariant__(dataset_id, lease):
+        """
+        The UUID of the dataset (key) must match the dataset UUID of
+        the Lease instance (value).
+        """
+        if dataset_id != lease.dataset_id:
+            return (False, "dataset_id {} does not match lease {}".format(
+                dataset_id, lease.dataset_id
+            ))
+        return (True, "")
+
+    def _check_lease(self, dataset_id, node_id):
+        """
+        Check if a lease for a given dataset is already held by a
+        node other than the one given and raise an error if it is.
+
+        :param UUID dataset_id: The dataset to check.
+        :param UUID node_uuid: The node that should hold a lease
+            on the given dataset.
+        """
+        if dataset_id in self and self[dataset_id].node_id != node_id:
+            raise ValueError("Lease already held by another node")
+
+    def acquire(self, now, dataset_id, node_id, expires=None):
+        """
+        Acquire and renew a lease.
+
+        :param datetime now: The current date/time.
+        :param UUID dataset_id: The dataset on which to acquire a lease.
+        :param UUID node_uuid: The node which will hold this lease.
+        :param int expires: The number of seconds from ``now`` until the
+            lease expires.
+        :return: The updated ``Leases`` representation.
+        """
+        try:
+            self._check_lease(dataset_id, node_id)
+        except ValueError as e:
+            message = (
+                u"Cannot acquire lease " + unicode(dataset_id) + " for node "
+                + unicode(node_id) + ": " + e.message
+            )
+            raise LeaseAcquisitionError(message)
```

Comment:
I think this can be simplified a bit:

1. There doesn't seem to be a need for two special exception types, one `LeaseError` will do.
2. There's no need for it to subclass `ValueError`.
3. :question: The exception class itself can construct the message.

This means that `_check_lease` could instead have `raise LeaseError(dataset_id, node_id)`, and the `try`/`except` blocks can be removed from `acquire` and `release`.